kv: raft closed timestamp is no longer nullable

surahman opened the pull request:
The Raft protocol closed timestamp is no longer nullable.

This change does not require a per-range migration because other migrations (namely, AddRaftAppliedIndexTermMigration) have already run across all ranges since v21.1 when this field was initially added. Those migrations ensured that all ranges now have a non-nil RaftClosedTimestamp.

Closes #84104.
Jira issue: [CRDB-17462](https://cockroachlabs.atlassian.net/browse/CRDB-17462)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84211)
<!-- Reviewable:end -->

nvb reviewed (commented):
Hi @surahman, thanks for the contribution. This looks like it's on the right track.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-:-N6kM5XJ8cEPBT_wDwzV:b-1qo45x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @surahman)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kK7y91te468g4k7Ll):*
This change has no external facing impact, so it does not need a release note.

However, we should include a commit message. You can use the first sentence that's written here for that. The second sentence is about v21.1, which is an old release that's no longer relevant, so we can delete it.

In the commit message, we should also mention that "This change does not require a per-range migration because other migrations (namely, AddRaftAppliedIndexTermMigration) have already run across all ranges since v21.1 when this field was initially added. Those migrations ensured that all ranges now have a non-nil RaftClosedTimestamp."
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kJCeSDYqSbu64Og9t:-N6kJCeSDYqSbu64Og9u:b-t2rk3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8ab1b1499cc65bf1887fd4d863511db59283fe/pkg/kv/kvserver/stateloader/stateloader.go#L190)):*
> ```Go
> // by the range applied state key.
> //
> // TODO(andrei): raftClosedTimestamp is a pointer to avoid an allocation when
> ```

We can now remove this TODO.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kJEYW4nT2p97hUZcS:-N6kJEYW4nT2p97hUZcT:bzb3hqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8ab1b1499cc65bf1887fd4d863511db59283fe/pkg/kv/kvserver/stateloader/stateloader.go#L199)):*
> ```Go
> 	appliedIndex, leaseAppliedIndex, appliedIndexTerm uint64,
> 	newMS *enginepb.MVCCStats,
> 	raftClosedTimestamp *hlc.Timestamp,
> ```

This parameter should now be a value.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kJKpX4jmlNA6ZdeKB:-N6kJKpX4jmlNA6ZdeKC:b-poapi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8ab1b1499cc65bf1887fd4d863511db59283fe/pkg/kv/kvserver/stateloader/stateloader.go#L208)):*
> ```Go
> 	}
> 	if raftClosedTimestamp != nil && !raftClosedTimestamp.IsEmpty() {
> 		as.RaftClosedTimestamp = *raftClosedTimestamp
> ```

Once `raftClosedTimestamp` is a value, we can set this directly in the `RangeAppliedState` struct literal.
___
*[`pkg/storage/enginepb/mvcc3.proto` line 262 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kJoxZ1NpXgrNfEk7i:-N6kJoxZ1NpXgrNfEk7j:b5kr0pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8ab1b1499cc65bf1887fd4d863511db59283fe/pkg/storage/enginepb/mvcc3.proto#L262)):*
> ```protobuf
>   // this timestamp.
>   //
>   // Not nullable because we cannot write empty timestamp to disk as that
> ```

Let's delete this entire paragraph.


<!-- Sent from Reviewable.io -->

surahman commented:
Thanks for the notes @nvanbenschoten I have affected changes.

surahman commented:
Bazel CI failure seems to be unrelated and failing test is passing locally.

```bash
Failed
=== RUN   TestBackupRestoreNonCCL/drop_schema
    cumulative.go:53: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8707/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/schemachangerccl/schemachangerccl_test_/schemachangerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/drop_schema:1:
        setup [0 args]
        CREATE DATABASE db;
        CREATE SCHEMA db.sc;
        ----
        ...
        +database {0 0 db} -> 104
        +schema {104 0 public} -> 105
        +schema {104 0 sc} -> 106
    --- FAIL: TestBackupRestoreNonCCL/drop_schema (463.69s)
```

[TestBackupRestoreNonCCL/drop_schema](https://github.com/cockroachdb/cockroach/blob/91d765006ddf92eb087596e62ffa7cb59392f4b3/pkg/sql/schemachanger/testdata/end_to_end/drop_schema)

nvb reviewed (commented):
Getting close!

> Reviewed 1 of 1 files at r7, 3 of 3 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-:-N7JIo0f7HX813rbLpXP:bvpx97u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @surahman)

___
*[`-- commits` line 4 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N7JI36j8m9uDuSb5eIg):*
Let's put this at the bottom of the commit message, with an empty line between it and the rest of the description.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kJKpX4jmlNA6ZdeKB:-N7JIJYSCSAfr1M8sITu:b-i7dga6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8ab1b1499cc65bf1887fd4d863511db59283fe/pkg/kv/kvserver/stateloader/stateloader.go#L208)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Once `raftClosedTimestamp` is a value, we can set this directly in the `RangeAppliedState` struct literal.
</blockquote></details>

We can address this comment now.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 235 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N7JIQPa-OMXW-N5rqpV:-N7JIQPa-OMXW-N5rqpW:bqqoglq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab56b026f190f7501a4b92ce84070448b5e7b9dd/pkg/kv/kvserver/stateloader/stateloader.go#L235)):*
> ```Go
> // SetClosedTimestamp overwrites the closed timestamp.
> func (rsl StateLoader) SetClosedTimestamp(
> 	ctx context.Context, readWriter storage.ReadWriter, closedTS *hlc.Timestamp,
> ```

Let's make this a value as well.


<!-- Sent from Reviewable.io -->

nvb reviewed (changes requested):
> I am not sure if we can switch this from a pointer to a value

I think what you did was correct. We can keep `initClosedTS` as a pointer.

> Reviewed 1 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-:-N7M7gtm3v4m0BVRGKO5:b-grx04w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @surahman)

___
*[`-- commits` line 4 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N7JI36j8m9uDuSb5eIg):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's put this at the bottom of the commit message, with an empty line between it and the rest of the description.
</blockquote></details>

Sorry, this was in reference to the release note, not the first line of the description. We should move this back and then move the release note down. The commit description should come before the release note.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84211#-N6kJKpX4jmlNA6ZdeKB:-N7M7p2hEmAdj5L1m8oD:b-mozfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf8ab1b1499cc65bf1887fd4d863511db59283fe/pkg/kv/kvserver/stateloader/stateloader.go#L208)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We can address this comment now.
</blockquote></details>

This can now look like:
```go
*as = enginepb.RangeAppliedState{
    RaftAppliedIndex:     appliedIndex,
    LeaseAppliedIndex:    leaseAppliedIndex,
    RangeStats:           newMS.ToPersistentStats(),
    RaftClosedTimestamp:  raftClosedTimestamp,
    RaftAppliedIndexTerm: appliedIndexTerm,
}
```
And we can get rid of
```go
if !raftClosedTimestamp.IsEmpty() {
    as.RaftClosedTimestamp = raftClosedTimestamp
}
```


<!-- Sent from Reviewable.io -->

surahman committed:
kv: make raft closed timestamp non-nullable

Addresses a TODO.

The Raft protocol closed timestamp is no longer nullable. This change does not require a per-range
migration because other migrations (namely, AddRaftAppliedIndexTermMigration) have already run
across all ranges since v21.1 when this field was initially added. Those migrations ensured that
all ranges now have a non-nil RaftClosedTimestamp.

Release note: None

nvanbenschoten commented:
Thanks @surahman!

bors r=nvanbenschoten

surahman commented:
thank you @nvanbenschoten
